libs/wingc/src/lsp/hover.rs

```diff
@@ -27,45 +203,52 @@ pub unsafe extern "C" fn wingc_on_hover(ptr: u32, len: u32) -> u64 {
 pub fn on_hover<'a>(params: lsp_types::HoverParams) -> Option<Hover> {
 	FILES.with(|files| {
 		let files = files.borrow();
-		let files = files.read().unwrap();
 		let parse_result = files.get(&params.text_document_position_params.text_document.uri.clone());
 		let parse_result = parse_result.unwrap();
 
 		let position = params.text_document_position_params.position;
 
 		let root_scope = &parse_result.scope;
-		let root_env_ref = root_scope.env.borrow();
-		let root_env = root_env_ref.as_ref().unwrap();
+		let root_env = root_scope.env.borrow();
+		let root_env = root_env.as_ref().unwrap();
+
+		let mut hover_visitor = HoverVisitor::new(position);
+		hover_visitor.visit_scope(root_scope);
 
-		if let Some(symbol_context) = find_symbol(root_scope, &position) {
-			let symbol = symbol_context.1;
-			let expr = symbol_context.0.expression;
-			let scope = symbol_context.0.scope;
+		if let Some(symbol) = hover_visitor.found_symbol {
+			let scope = hover_visitor.current_scope.expect("All symbols must be in a scope");
+			let expr = hover_visitor.current_expr;
 			let symbol_name = &symbol.name;
-			let expression_type = expr.as_ref().and_then(|expr| {
+			let expression_type = expr.and_then(|expr| {
 				let t = expr.evaluated_type.borrow();
 				t.clone()
 			});
-			let reference = expr.as_ref().and_then(|expr| {
+			let reference = expr.and_then(|expr| {
 				if let ExprKind::Reference(reference) = &expr.kind {
 					Some(reference)
 				} else {
 					None
 				}
 			});
 			let span = if let Some(_) = reference {
-				&expr.unwrap().span
+				&expr.expect("Missing type for expression").span
 			} else {
 				&symbol.span
 			};
```

Comment:
What is the reason for using the parent Expr's span instead of the symbol's span? (maybe consider adding a comment)

Reply:
<img width="123" alt="image" src="https://user-images.githubusercontent.com/1237390/220984151-a909a9f8-1cb6-470c-9728-baed2b214c8f.png">
This is maybe controversial since TS and rust don't do this, but I kinda like that this highlights the entire relevant part of a reference rather than just the field.

I'll add a comment